### ISSUE TITLE
CFP: Being able to specify multiple CIDR ranges in `ipv4-native-routing-cidr` field

### ISSUE THREAD

**Issue by DockToFuture** [cilium/cilium#24000]
## Cilium Feature Proposal
        
Thanks for taking time to make a feature request for Cilium! If you have usage questions, please try the [slack channel](http://slack.cilium.io/) and see the [FAQ](https://github.com/cilium/cilium/issues?utf8=%E2%9C%93&q=is:issue+label:kind/question+) first.

**Is your feature request related to a problem?**
I've set up a kubernetes cluster with cilium which is run in native routing mode. I want to masquerade all traffic which is not part of the pod CIDR range and node CIDR range. In the node CIDR range I want to masquerade all traffic which is going to two IPs (upstream dns server). For that I've created a separate masquerading rule in the cilium `CILIUM_POST_nat` chain in the NAT table.

Cilium is run with the following options to masquerade traffic:
```
enable-ipv4-masquerade: true
ipv4-native-routing-cidr: <nodeCIDR>
```

Unfortunately I can only specify one CIDR range in the `ipv4-native-routing-cidr` field. I would like two specify two ranges for which masquerading is disabled (`<nodeCIDR> and <podCIDR>` range).

In the cilium documentation is written, that I can use the ip masq agent to specify multiple non-masquerading CIDRs. 
However it requires bpf-masquerading, which does not allow me to configure masquerading rules in the  `CILIUM_POST_nat` chain. Furthermore the ip masq agent requires `enable-bpf-masquerade=true` which requires `enable-node-port=true` which is automatically disabled because we are running without kube-proxy (`"Auto-disabling \"enable-node-port\", \"enable-external-ips\", \"bpf-lb-sock\", \"enable-host-port\" features and falling back to \"enable-host-legacy-routing\"" subsys=daemon`) . It's is also not clear what is the target picture of cilium here, I thought the goal was to replace kube-proxy but then I would expect, that I'm capable of using the ip masq agent without kube-proxy which is currently not possible.

**Describe the feature you'd like**

I would like to be able to specify multiple CIDR ranges in the `ipv4-native-routing-cidr` field for which masquerading is dropped.

**(Optional) Describe your proposed solution**

Make a string array out of `ipv4-native-routing-cidr`

**Additional**

If there is a better way to deal with that problem I'm just not aware of tell me. :)

**Comment by hujun-open** [comment]
have same issue for ipv6-native-routing-cidr

**Comment by erikschul** [comment]
@joestringer Sorry to ping you specifically, but I'm unsure who else to write to.
This seems to be a common problem (18 thumbs up), but since it's closed, I suspect no one is aware of it. 
Would it make sense to reopen it?

**Comment by joestringer** [comment]
Sure, I can reopen. I've been thinking a little bit about this in the background recently anyway, so I've also started some development discussion about improving usability in this area over at this issue https://github.com/cilium/cilium/discussions/36335 .

As far as I understand, there are two possible solutions to configure multiple non-masquerade CIDRs:
1. Use upstream ipmasq-agent which is based on iptables
2. Use Cilium's [bpf-based ipmasq agent implementation](https://docs.cilium.io/en/stable/network/concepts/masquerading) which is compatible with the same configmaps

It looks like the limitation that requires bpf nodeport for bpf-based masquerade is still present (tracked in https://github.com/cilium/cilium/issues/13732).

**Comment by joestringer** [comment]
> In the node CIDR range I want to masquerade all traffic which is going to two IPs (upstream dns server).

Do I follow right that for this use case you want Pod traffic destined to specific Node CIDR ranges to be masqueraded even though the traffic is actually routable on the node CIDR range? Abstractly I would expect that Cilium should be able to integrate with existing Linux networking stack configuration by setting up Cilium to hand traffic to the Linux networking stack for further processing, where you would then create any related iptables or route configuration. Any traffic that is not destined towards a Cilium-managed destination should then go to the Linux stack and work with any existing rules you've configured. I was under the impression that this is generally true of Cilium in a default configuration, but it probably needs more debugging to understand why that might not be the case / not working for you. Some combination of `hubble observe`, `cilium monitor` (with debug and no monitor aggregation) `tcpdump` and perhaps pwru could help to trace how the packets are being forwarded to try to understand further.

If this specific aspect of the problem is a common use case, then it may make sense for Cilium to have a first-party configuration for this particular problem. I'm not sure I quite understand this specific use case well enough yet to propose a solution.

> It's is also not clear what is the target picture of cilium here, I thought the goal was to replace kube-proxy but then I would expect, that I'm capable of using the ip masq agent without kube-proxy which is currently not possible.

I think you understand the goals right, we aim to move everything into eBPF but it is a process over time, driven by community use cases and contributions. Over time we discover more and more ways that users are using Cilium so it can be a gradual process to ensure that the various use cases are well-covered by the implementation.

**Comment by ross-c-swyftx** [comment]
I have another use case for a feature that extends the IP ranges specified for native routing. 

We are in the middle of deploying VPC lattice for inter-service communication between services in separate EKS clusters. Some clusters are legacy and aren't Cilium enabled, so no clustermesh. 

The lattice network uses IP addresses in the range `169.254.171.0/24`, all our clusters have IP addresses in the range `10.0.0.0/8`. When a service gets allocated to a secondary ENI IP range (as opposed to the primary), traffic cannot reach the service due to asymmetric routing (traffic from secondary ENI headed to lattice range exits through the primary ENI). Cilium appears to only sets up routes for the 10.0.0.0/8 range on secondary ENIs. If this feature was implemented we could add the VPC lattice range and it appears Cilium, by default, would then add that range to the secondary ENI lookup table.

**Comment by superbrothers** [comment]
We are currently using Native Routing mode. We are considering using node-ipam-controller to expand the Pod CIDR, but we are unable to prepare a new continuous CIDR. In such cases, we believe we need to be able to specify multiple CIDR ranges for the ipv4-native-routing-cidr field.

https://github.com/kubernetes-sigs/node-ipam-controller
https://docs.cilium.io/en/stable/network/concepts/routing/#native-routing

**Comment by netops2devops** [comment]
@superbrothers Consider looking into Multi Pool IPAM specially if you are using Native Routing.

**Comment by superbrothers** [comment]
> Consider looking into Multi Pool IPAM specially if you are using Native Routing.

Thank you for the information. Is this the documentation for "Multi Pool IPAM"? 

https://docs.cilium.io/en/latest/network/concepts/ipam/multi-pool/

**Comment by netops2devops** [comment]
Yes! That's the one. The Cilium team flushed out majority of the bugs in 1.18 release so in its current form it's quite stable. In 1.19.0 release it will "officially" be moved from Beta to Stable in the docs.
